File: code/jython/apksigner/utils/simple_weak_encryption.py
#  Copyright (C) 2012 Hai Bison
#
#  See the file LICENSE at the root directory of this project for copying
#  permission.

'''
This is the **simple-and-weak** encryption utilities.
'''

import array
import base64
import sys

from java.security.spec import KeySpec

from javax.crypto import Cipher, SecretKey, SecretKeyFactory
from javax.crypto.spec import IvParameterSpec, PBEKeySpec, SecretKeySpec

_TRANSFORMATION = "AES/CBC/PKCS5Padding"
_SECRET_KEY_FACTORY_ALGORITHM = "PBKDF2WithHmacSHA1"
_SECRET_KEY_SPEC_ALGORITHM = "AES"

'''
Only ``128`` bits. If this is ``256``, some JVMs must need extra tools to be
installed. Poor Java :|
'''
_KEY_LEN = 128
_ITERATION_COUNT = 2**16
_SEPARATOR = '\t'

_UTF8 = 'UTF-8'

def encrypt_to_base64(password, data):
    ''' Encrypts ``data`` (String) by ``password`` (char[]).
    '''

    cipher = Cipher.getInstance(_TRANSFORMATION)
    cipher.init(Cipher.ENCRYPT_MODE, gen_key(password))

    data = cipher.doFinal(data.encode(_UTF8))

    return '{}{}{}'.format(base64.b64encode(cipher.getIV().tostring()),
                           _SEPARATOR, base64.b64encode(data.tostring()))
    #.encrypt_to_base64()

def decrypt_from_base64(password, data):
    ''' Decrypts an encrypted string (``data`` -- String) by ``password``
        (char[]).

        Returns:
            the decrypted string, or ``None``.
    '''

    cipher = Cipher.getInstance(_TRANSFORMATION)
    idx = data.find(_SEPARATOR)
    if idx <= 0: return

    try:
        cipher.init(
            Cipher.DECRYPT_MODE,
            gen_key(password),
            IvParameterSpec(array.array('b', base64.b64decode(data[:idx]))))
    except:
        print(' ! Error: {}'.format(sys.exc_info()[1]))
        # Ignore it
        return

    try:
        data = array.array('b', base64.b64decode(data[idx+1:]))
        return cipher.doFinal(data).tostring()
    except:
        print(' ! Error: {}'.format(sys.exc_info()[1]))
        # Ignore it
    #.decrypt_from_base64()

def gen_key(password):
    ''' Generates secret key.

        Parameters:

        :password (char[]):
            the password.
    '''

    factory = SecretKeyFactory.getInstance(_SECRET_KEY_FACTORY_ALGORITHM)
    spec = PBEKeySpec(password, str(password).encode(_UTF8),
                      _ITERATION_COUNT, _KEY_LEN)
    tmp = factory.generateSecret(spec)

    return SecretKeySpec(tmp.getEncoded(), _SECRET_KEY_SPEC_ALGORITHM)
    #.gen_key()

File: code/scala/src/main/scala/group/pals/desktop/app/apksigner/utils/SimpleWeakEncryption.scala
/*
 *    Copyright (C) 2012 Hai Bison
 *
 *    See the file LICENSE at the root directory of this project for copying
 *    permission.
 */

package group.pals.desktop.app.apksigner.utils

import java.security.Key

import android.util.Base64
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * The <b>simple-and-weak</b> encryption utilities.
 *
 * @author Hai Bison.
 *
 */
object SimpleWeakEncryption {

    private lazy final val TRANSFORMATION = "AES/CBC/PKCS5Padding"
    private lazy final val SECRET_KEY_FACTORY_ALGORITHM = "PBKDF2WithHmacSHA1"
    private lazy final val SECRET_KEY_SPEC_ALGORITHM = "AES"

    /**
     * Only {@code 128} bits. If this is {@code 256}, some JVMs must need extra
     * tools to be installed. Poor Java :|
     */
    private lazy final val KEY_LEN = 128
    private lazy final val ITERATION_COUNT = Math.pow(2, 16).asInstanceOf[Int]
    private lazy final val SEPARATOR = "\t"
    private lazy final val DEFAULT_BASE64_FLAGS = Base64.NO_WRAP

    lazy final val UTF8 = "UTF-8"

    /**
     * Encrypts {@code data} by {@code key}.
     *
     * @param password
     *            the secret key.
     * @param data
     *            the data.
     * @return the encrypted data.
     */
    def encrypt(password: Array[Char], data: String): String = {
        var cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, genKey(password))

        Base64.encodeToString(cipher.getIV(), DEFAULT_BASE64_FLAGS) +
            SEPARATOR +
            Base64.encodeToString(
                cipher.doFinal(data.getBytes(UTF8)), DEFAULT_BASE64_FLAGS)
    } // encrypt()

    /**
     * Decrypts an encrypted string ({@code data}) by {@code key}.
     *
     * @param password
     *            the password.
     * @param data
     *            the data.
     * @return the decrypted string.
     */
    def decrypt(password: Array[Char], data: String): String = {
        val iSeparator = data.indexOf(SEPARATOR)
        var cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.DECRYPT_MODE, genKey(password),
            new IvParameterSpec(
                Base64.decode(data.substring(0, iSeparator), DEFAULT_BASE64_FLAGS)))
        new String(
            cipher.doFinal(Base64.decode(
                data.substring(iSeparator + 1), DEFAULT_BASE64_FLAGS)),
            UTF8)
    } // decrypt()

    /**
     * Generates secret key.
     *
     * @param password
     *            the password.
     * @return the secret key.
     */
    private def genKey(password: Array[Char]): Key = {
        var factory = SecretKeyFactory.getInstance(SECRET_KEY_FACTORY_ALGORITHM)
        var spec = new PBEKeySpec(
            password, new String(password).getBytes(UTF8), ITERATION_COUNT,
            KEY_LEN)
        new SecretKeySpec(
            factory.generateSecret(spec).getEncoded(),
            SECRET_KEY_SPEC_ALGORITHM)
    } // genKey()

}
